src/hotspot/share/services/rsswatch.cpp

```diff
@@ -0,0 +1,132 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+
+#include "logging/log.hpp"
+#include "runtime/java.hpp"
+#include "runtime/os.hpp"
+#include "runtime/task.hpp"
+#include "services/rsswatch.hpp"
+#include "utilities/parseInteger.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/globalDefinitions.hpp"
+
+static void check_rss(size_t limit) {
+  const size_t rss = os::get_rss();
+  log_trace(os, rss)("Rss=%zu", rss);
+  if (rss >= limit) {
+    fatal("Resident Set Size (%zu bytes) reached RssLimit (%zu bytes).", rss, limit);
+  }
+}
+
+class RssAbsoluteLimitTask : public PeriodicTask {
+  const size_t _limit;
+public:
+  RssAbsoluteLimitTask(size_t limit, unsigned interval_ms) :
+    PeriodicTask(interval_ms), _limit(limit) {
+    log_info(os, rss)("RssWatcher task: interval=%ums, limit=" PROPERFMT,
+                      interval(), PROPERFMTARGS(_limit));
+  }
+  void task() override {
+    check_rss(_limit);
+  }
+};
+
+class RssRelativeLimitTask : public PeriodicTask {
+  const double _percent;
+  size_t _limit;
+
+  void update_limit() {
+    const size_t limit_100 = os::physical_memory();
+    const size_t limit_x = (size_t)(((double)limit_100 * _percent) / 100.0f);
+    if (limit_x != _limit) { // limit changed?
+      _limit = limit_x;
+      log_info(os, rss)("Setting RssWatcher limit to %zu bytes (%.2f%% of total memory of %zu bytes)",
+                        limit_x, _percent, limit_100);
+    }
+  }
+
+public:
+
+  RssRelativeLimitTask(double limit_percent, unsigned interval_ms) :
+    PeriodicTask(interval_ms), _percent(limit_percent), _limit(0) {
+    log_info(os, rss)("RssWatcher task: interval=%ums, "
+                      "limit=%.2f%% of total memory", interval(), _percent);
+  }
+
+  void task() override {
+    update_limit();
+    check_rss(_limit);
+  }
+};
+
+static bool parse_percentage(const char* s, const char** tail, double* percentage) {
+  double v = 0;
+  char sign;
+  int chars_read = 0;
+  if (sscanf(s, "%lf%c%n", &v, &sign, &chars_read) >= 2 && sign == '%') {
+    if (v > 100.0) {
+      vm_exit_during_initialization("Failed to parse RssLimit", "Not a valid percentage");
+    }
+    *percentage = v;
+    *tail = s + chars_read;
+    return true;
+  }
+  return false;
+}
+
+void RssWatcher::initialize(const char* limit_option) {
+  assert(limit_option != nullptr, "Invalid argument");
+
+  bool is_absolute = true;
+  size_t limit = 0;
+  double percentage = 0;
+  const char* s = limit_option;
+
+  if (parse_percentage(s, &s, &percentage)) {
+    is_absolute = false;
+  } else {
+    if (!parse_integer(s, (char**)&s, &limit) || limit == 0) {
+      vm_exit_during_initialization("Failed to parse RssLimit", "Not a valid limit size");
```

Comment:
You specified that zero turned the feature off

Reply:
Ah, good catch.